pyrokinetics/gk_code/GKOutputReaderCGYRO.py

```diff
@@ -407,33 +483,101 @@ def _set_eigenfunctions(
         This should be called after _set_fields, and is only valid for linear runs.
         """
 
-        # Use default method to calculate growth/freq if possible
-        all_ballooning = data.ntheta_plot == data.ntheta_grid
-        fields_contains_nan = np.any(np.isnan(data["fields"].data))
-        fields_exist = [(f"field_{f}" in raw_data) for f in data["field"].data]
-        if all_ballooning and np.all(fields_exist) and not fields_contains_nan:
-            data = GKOutputReader._set_eigenfunctions(data, raw_data, gk_input)
-            return data
-
         raw_eig_data = [
-            raw_data.get(f"eigenfunctions_{f}", None) for f in data["field"].data
+            raw_data.get(f"eigenfunctions_{f}", None) for f in coords["field"]
         ]
-        raw_shape = [2, data.ntheta, data.ntime]
+
+        ntime = len(coords["time"])
+        ntheta = len(coords["theta"])
+        nkx = len(coords["kx"])
+        nky = len(coords["ky"])
+
+        raw_shape = [2, ntheta, nkx, nky, ntime]
 
         # FIXME Currently using kx and ky for compatibility with GS2 results, but
         #       these coordinates are not used. Should we remove these coordinates?
-        coords = ["kx", "ky", "field", "theta", "time"]
-        eigenfunctions = np.empty([data.dims[coord] for coord in coords], dtype=complex)
+        coord_names = ["field", "theta", "kx", "ky", "time"]
+        eigenfunctions = np.empty(
+            [len(coords[coord_name]) for coord_name in coord_names], dtype=complex
+        )
 
         # Loop through all fields and add eigenfunction if it exists
         for ifield, raw_eigenfunction in enumerate(raw_eig_data):
             if raw_eigenfunction is not None:
                 eigenfunction = raw_eigenfunction[: np.prod(raw_shape)].reshape(
                     raw_shape, order="F"
                 )
-                eigenfunctions[:, :, ifield, :, :] = (
-                    eigenfunction[0] + 1j * eigenfunction[1]
-                )
+                eigenfunctions[ifield, ...] = eigenfunction[0] + 1j * eigenfunction[1]
 
-        data["eigenfunctions"] = (coords, eigenfunctions)
-        return data
+        square_fields = np.sum(np.abs(eigenfunctions) ** 2, axis=0)
+        field_amplitude = np.sqrt(np.trapz(square_fields, coords["theta"], axis=0)) / (
+            2 * np.pi
+        )
+        result = eigenfunctions / field_amplitude
+
+        return result
+
+    @staticmethod
+    def to_netcdf(self, *args, **kwargs) -> None:
```

Comment:
I'm not sure why the `to_netcdf` and `from_netcdf` functions are defined here (similarly for other reader classes). It looks to me like they're just handling complex data for netcdf IO, which is something we should probably be doing in `dataset_wrapper`  by default. Could we just move these lines there? If we only want this behaviour in `GKOutput`, we could define these functions there instead (but also use `super()` to avoid repeat code if possible).

(staticmethods also shouldn't have a `self` arg, as in this context `self` will just be a standard function argument that you can pass in, and it doesn't operate on a particular instance of the class. I think `@staticmethod` just needs to be removed in this case)

Reply:
Not sure why I did that originally but I've moved it to `dataset_wrapper` now